src/main/java/util/RefreshTimer.java

```diff
@@ -0,0 +1,105 @@
+     * This amount is then deducted from the actual apiQuota to ensure the algorithm reserves this amount.
+     * The calculation has 2 steps.
+     * The first step, the offset amount(apiQuota - APIQUOTA_BUFFER) is divided by the amount
+     * of api used during the last refresh. This amount corresponds to the estimated no of refresh allowed to use till
+     * the next apiQuota renewal.
+     * Second, the refresh rate is calculated by dividing the remainingTimeInMin by the estimated no of refresh allowed.
+     *
+     * For some case where the refresh rate is equal to the remainingTimeInMin till next apiQuota renewal. BUFFER_TIME
+     * is added to ensure that the next refresh happens after the apiQuota renewal.
+     *
+     * @param apiQuota The remaining allowed api request until the next api request allowance renewal.
+     * @param remainingTimeInMin The remaining time left until the next api request allowance renewal.
+     * @param lastApiCallsUsed The amount of api used in the last api pull.
+     * @param bufferTimeInMin The amount of time in min to be added to the refreshTime in cases when computed
+     *                        refresh period equals to remainingTimeInMin. Recommended value : RefreshTimer.BUFFER_TIME
+     * @param defaultRefreshPeriodInMin : The default refresh period in min that will be used when input arguments
+     *                                    value is not valid for calculation of refresh rate. In those cases,
+     *                                    the default value is deemed to be the most proper refresh rate.
+     *                                    Recommended value : RefreshTimer.DEFAULT_REFRESH_PERIOD_IN_MIN
+     * @return Computed refresh period.
+     */
+    public static long computeRefreshTimerPeriod(int apiQuota, long remainingTimeInMin, int lastApiCallsUsed,
+                                                 int bufferTimeInMin, int defaultRefreshPeriodInMin) {
+
+        assert apiQuota >= 0 && remainingTimeInMin >= 0 && lastApiCallsUsed >= 0 && bufferTimeInMin >= 0
+                && defaultRefreshPeriodInMin > 0;
+
+        long refreshTimeInMin;
+
+        boolean isDuringAppInit = apiQuota != 0 && lastApiCallsUsed == 0;
+        if (isDuringAppInit || remainingTimeInMin == 0) {
+            refreshTimeInMin = defaultRefreshPeriodInMin;
+            return refreshTimeInMin;
+        }
+
+        if (isQuotaInsufficient(apiQuota, lastApiCallsUsed)) {
+            refreshTimeInMin = remainingTimeInMin + bufferTimeInMin;
+            return refreshTimeInMin;
+        }
+
+        double noOfRefreshAllowed = computeNoOfRefreshAllowed(apiQuota, lastApiCallsUsed);
+        refreshTimeInMin = (long) Math.ceil(remainingTimeInMin / noOfRefreshAllowed);
+
+        if (refreshTimeInMin == remainingTimeInMin) {
+            refreshTimeInMin = refreshTimeInMin + bufferTimeInMin;
+        }
+        return refreshTimeInMin;
```

Comment:
At least give it a try to see if that makes the code easier to understand. I have a feeling that the calculation can be simpler but the way to simplify is harder to see because of the extra abstraction. I may be wrong.